imxrt gptp improvements

ainguraXmarquiegui opened the pull request:
gptp implementation for NXP imxrt platform is broken and buggy. This merge request won't fix all problems with gptp, but it's a significant improvement because it fixes a few key bugs in the ethernet driver.

This code base has been tested on imxrt1050 and imxrt1060 platforms. Synchronization is still not as precise or as stable as it should be, but it's a clear improvement from what is currently on mainline, since the ethernet driver doesn't crash anymore.

All these modifications have been contributed during the discussion of this issue:

https://github.com/zephyrproject-rtos/zephyr/issues/33747

ainguraXmarquiegui committed:
net: gptp: convert clock sync ratio from float to double

Using clock sync ratio as double instead of float improves
synchronization smoothness

Signed-off-by: Xabier Marquiegui <xmarquiegui@ainguraiiot.com>

ainguraXmarquiegui commented:
I think some of the commits in this pull request don't meet the zerphyr contributing guidelines.

@hakehuang , how should we proceed to clean this up?

hakehuang commented:
> I think some of the commits in this pull request don't meet the zerphyr contributing guidelines.
> 
> @hakehuang , how should we proceed to clean this up?

first merge all the commits into one, and then seperate them iin function commits

erwango commented:
^^@alsrgv FYI (ptp_clock.h impact)

hakehuang commented:
@ainguraXmarquiegui , the second commit need separate to pinmux, ethdriver, board config and gptp stack changes, also for ethernet change, please add `Co-authored-by: Seb Laveze <sebastien.laveze@nxp.com>` Thanks.

ainguraXmarquiegui commented:
@hakehuang . I have done the necessary rebase and cleanup, but I'm not sure I will be able to easily separate the code and write meaningful commits. Can you do the commit sepparation since you know the changes better than I do?

ainguraXmarquiegui committed:
samples: net: gptp: add NXP RT series config support

Add NXP RT samples config support for gptp sample

Signed-off-by: Hake Huang <hake.huang@oss.nxp.com>
Co-authored-by: Seb Laveze <sebastien.laveze@nxp.com>
Signed-off-by: Xabier Marquiegui <xmarquiegui@ainguraiiot.com>

ainguraXmarquiegui committed:
boards: mimxrt: enable pps output

Enable pps output for mimxrt1050 and mimxrt1060 based boards for gptp
sample application validation.

Signed-off-by: Hake Huang <hake.huang@oss.nxp.com>
Co-authored-by: Seb Laveze <sebastien.laveze@nxp.com>
Signed-off-by: Xabier Marquiegui <xmarquiegui@ainguraiiot.com>

ainguraXmarquiegui commented:
@hakehuang , forget about my last post. Please review my attempt at doing what you requested. Is this separation what you wanted?

dleach02 commented:
For traceability, can you enter in issues for the specific issues you are fixing with this PR

ainguraXmarquiegui commented:
All these modifications have been contributed during the discussion of this issue:

https://github.com/zephyrproject-rtos/zephyr/issues/33747

danieldegrasse reviewed (approved):
I don't have a full setup, but running `samples/net/gptp` on an RT1060, using linuxptp and the configuration file provided in [this comment](https://github.com/zephyrproject-rtos/zephyr/issues/33747#issuecomment-998981509), gptp appears to work correctly, and `net gptp` reports that the current state is `RECEIVING_SYNC`. I do see the following console message after a few seconds, though:
```
<wrn> net_gptp: Not AS capable: 4294967295 ns > 1000000 ns
```

danieldegrasse commented on drivers/ethernet/eth_mcux.c:
```diff
@@ -930,6 +944,46 @@ static void eth_callback(ENET_Type *base, enet_handle_t *handle,
 	}
 }
 
+static void eth_rx_thread(void *arg1, void *unused1, void *unused2)
```
Would it be possible to split this commit into a few separate ones?
For example, something like this:
- commit moving Ethernet RX/TX handlers to threads
- commit that fixes the `ENET_SendFrame error: 4004` issue (if this was a specific, smaller changeset)
- commit with remaining PTP specific changes

Feel free to divide up this commit however you see fit, but I think it would make this PR much more approachable if the changes to the Ethernet driver were split across commits.

ainguraXmarquiegui replied:
Thank you for your suggestion @danieldegrasse . I have done a first attempt at splitting the commit. Have a look, and let me know if you like the new proposal better.

mmahadevan108 commented on boards/arm/mimxrt1060_evk/pinmux.c:
```diff
@@ -219,6 +219,18 @@ static int mimxrt1060_evk_init(const struct device *dev)
+	IOMUXC_SetPinMux(IOMUXC_GPIO_AD_B1_03_ENET_1588_EVENT2_IN, 0U);
+	/* GPIO_AD_B1_03 PAD functional properties */
+	IOMUXC_SetPinConfig(IOMUXC_GPIO_AD_B1_03_ENET_1588_EVENT2_IN, 0xB0E9);
+#endif
```
Please update to the documentation to add these pins

ainguraXmarquiegui replied:
@mmahadevan108 , please, expand your comment. I don't understand what you mean.

mmahadevan108 replied:
Looks like new pins have been defined here. Please add it to the `index.rst` file in the doc folder

ainguraXmarquiegui replied:
OK, I have just force pushed the modified index.rst files. Please, review my attempt and let me know if I should do any additional modifications. If you like what I uploaded just mark the conversation as resolved. Thanks!

mmahadevan108 commented on samples/net/gptp/boards/mimxrt1020_evk.conf:
```diff
@@ -0,0 +1,2 @@
+CONFIG_PTP_CLOCK_MCUX=y
```
Has this been tested on `RT1020`

ainguraXmarquiegui replied:
I did some tests back in the day. gPTP is not working as expected in any imxrt model, including the imxrt1020, imxrt1050 and imxrt1060. According to the chip documentation, the imxrt1020 does have IEEE1588 hardware support. This line only enables the feature in the gptp sample application for this model.

ainguraXmarquiegui committed:
driver: eth_mcux: realtime optimizations

mcux enet driver updates:
a) add mutex to access the ptp timer variables
b) use thread for rx(cooperative) andfor tx cleanup(cooperative)
c) use a thread for ptp inner clock updater(pre-empty)

This patch fixes the following issue:

```
ENET_SendFrame error: 4004
```

Signed-off-by: Hake Huang <hake.huang@oss.nxp.com>
Co-authored-by: Seb Laveze <sebastien.laveze@nxp.com>
Signed-off-by: Xabier Marquiegui <xmarquiegui@ainguraiiot.com>

ainguraXmarquiegui committed:
driver: eth_mcux: gptp: enable pps output

Enable pps output for gptp accuracy measurements.

Signed-off-by: Hake Huang <hake.huang@oss.nxp.com>
Signed-off-by: Xabier Marquiegui <xmarquiegui@ainguraiiot.com>

ainguraXmarquiegui committed:
driver: eth_mcux: gptp: limit rate_adjust range

Limit gptp range_adjust range for improved performance.

Signed-off-by: Hake Huang <hake.huang@oss.nxp.com>
Signed-off-by: Xabier Marquiegui <xmarquiegui@ainguraiiot.com>

alsrgv commented:
Thanks for looping me in, love the change of `rate` to double!

jukkar commented on samples/net/gptp/src/gptp.c:
```diff
@@ -75,9 +84,19 @@ void init_testing(void)
 	if (run_duration == 0) {
-		return;
+		/* USER BEGIN MAIN.C*/
+		while (1) {
+
+			k_msleep(200); /* sleep time in ms */
+		}
```
Why do we need this loop?

ainguraXmarquiegui replied:
@hakehuang , would you like to comment on this?

ainguraXmarquiegui replied:
I have just reviewed this, and I can answer this myself. It was my bad. That code snippet was there as part of a modification that I decided should not be part of this pull request. I removed the critical stuff, but I mistakenly left this there.

I have now removed it, so I'll mark this conversation as resolved. Thank you for pointing it out.

jukkar commented on subsys/net/l2/ethernet/gptp/gptp_md.c:
```diff
@@ -583,6 +583,12 @@ static void gptp_md_pdelay_req_state_machine(int port)
 		state->state = GPTP_PDELAY_REQ_NOT_ENABLED;
```
You are mixing core gptp change with this sample change. Please create a separate commit for the core gptp change.

jukkar commented on subsys/net/l2/ethernet/gptp/gptp_md.c:
```diff
@@ -583,6 +583,12 @@ static void gptp_md_pdelay_req_state_machine(int port)
 		state->state = GPTP_PDELAY_REQ_NOT_ENABLED;
 	}
 
+	while (1) {
+		if (net_if_flag_is_set(GPTP_PORT_IFACE(port), NET_IF_UP)) {
+			break;
+		}
+		k_sleep(K_MSEC(1000));
+	}
```
I am not sure this is a good idea, this will block the calling thread. Why can't we just return if the interface is not up?

ainguraXmarquiegui replied:
@hakehuang , can you provide your insights on this?

ainguraXmarquiegui replied:
I have just pushed an alternative proposal for that. I basically removed that while/sleep combo, and replaced it with this:

https://github.com/zephyrproject-rtos/zephyr/pull/42216/commits/44ea9cd76a54cfd3b0fecbcd4c32fbc900c8160e

Does it look good?

jukkar commented on drivers/ethernet/eth_mcux.c:
```diff
@@ -157,22 +157,35 @@ struct eth_context {
 	struct k_work phy_work;
 	struct k_work_delayable delayed_phy_work;
+
+	K_KERNEL_STACK_MEMBER(rx_thread_stack, 1600);
```
Please avoid magic constants, use a `#define` instead here for the size.

ainguraXmarquiegui replied:
Thank you for your suggestion. I have now used a define for this. Please, mark the conversation as resolved if you like my new proposal.

jukkar commented on drivers/ethernet/eth_e1000.c:
```diff
@@ -380,7 +380,7 @@ static int ptp_clock_e1000_adjust(const struct device *dev, int increment)
 	return 0;
```
No objection to the float -> double change. Please change the commit subject to
`net: gptp: clock sync ratio as double, not float`

Please also add information about this change to release notes in API changes section.

ainguraXmarquiegui replied:
Done. I hope I did it all right. Please, mark this conversation as resolved if you do, or give me some more input if I did something wrong.

Thank you very much for all your suggestions, @jukkar .

ainguraXmarquiegui commented on samples/net/gptp/src/gptp.c:
```diff
@@ -13,14 +13,23 @@ LOG_MODULE_DECLARE(net_gptp_sample);
 #include "ethernet/gptp/gptp_messages.h"
 #include "ethernet/gptp/gptp_data_set.h"
 
+/*USER BEGIN INCLUDES*/
```
Done, thank you very much for your suggestion.

ainguraXmarquiegui committed:
net: gptp: gptp port role disabled on link down

Set GPTP port role to disabled if port link is down.

Signed-off-by: Xabier Marquiegui <xmarquiegui@ainguraiiot.com>

ainguraXmarquiegui committed:
samples: net: gptp: Enable deferred log mode

In applications where timing is critical it seems like enabling
deferred log mode could be desirable.

Signed-off-by: Hake Huang <hake.huang@oss.nxp.com>
Co-authored-by: Seb Laveze <sebastien.laveze@nxp.com>
Signed-off-by: Xabier Marquiegui <xmarquiegui@ainguraiiot.com>

jukkar reviewed (commented):
Looks good, minor nit about commit subject.

ainguraXmarquiegui commented on subsys/net/l2/ethernet/gptp/gptp.c:
```diff
@@ -521,18 +521,23 @@ static void gptp_state_machine(void)
 	for (port = GPTP_PORT_START; port < GPTP_PORT_END; port++) {
```
Done. Fixed. Thanks!

jukkar reviewed (approved):
LGTM

dleach02 commented on doc/releases/release-notes-3.0.rst:
```diff
@@ -278,6 +278,7 @@ Networking
 
 * Misc:
 
+  * gptp: clock sync ratio as double, not float
```
If you want this PR to be part of the 3.0 release, you need to add a milestone and bring it to the attention of the release manager.

dkalowsk replied:
Thanks.  I would suggest changing this entry to `grtp: convert clock sync ratio from float to double` as to me the current text is confusing.

dleach02 replied:
@dkalowsk, I was specially recommending to @ainguraXmarquiegui, that he separate the release notes change to a separate PR that is pending based on whether this PR gets merged.

ainguraXmarquiegui replied:
Thank you @dleach02 . I have never used the Milestone feature on github.

If I create a milestone in my forked repo, I can't assign this pull request to that milestone, and I don't have permissions to create milestones here.

How does this typically work?

Should I remove the release notes commit from this PR and move it to a new PR so that someone with permissions can create a new milestone on this repo?

Or should I create the milestone along with two new PR on my own fork (if that's even possible)?

dleach02 replied:
the milestone selection is for this particular PR. It just needs the 3.0 selected. I can set the milestone if that is what you want.

I do want you to separate out the release note changes into a standalone PR that is pending this one being merged.

ainguraXmarquiegui replied:
OK, I removed the commit from here and created https://github.com/zephyrproject-rtos/zephyr/pull/42624 with only the release notes commit.

Let me know if this is good enough.
